ASv2 Contract kit wrappers

Jesse-Sawa opened the pull request.

Jesse-Sawa committed:
add contractkit wrappers for FederatedAttestations & OdisPayments (#9915)

* add contract kit wrappers for federated attestations

* ad odisPayments contract wrapper

Jesse-Sawa committed:
Add Escrow contract wrapper changes (#9917)

* Add Escrow Wrapper types and comments
Add ASv2 changes to Escrow Wrapper
Add Escrow Wrapper tests

* replace date with hardcoded timestamp

* remove string casting

Jesse-Sawa commented:
[Merge lint fixes first](https://github.com/celo-org/celo-monorepo/pull/9919)

Jesse-Sawa committed:
Fix ts linting errors (#9919)

fix ts-linting errors for asv2 contract kit wrapper additions

alecps commented on packages/sdk/contractkit/src/wrappers/OdisPayments.ts:
```diff
@@ -0,0 +1,26 @@
+import { CeloTransactionObject } from '@celo/connect'
```
There's already an implementation of this wrapper in the odisCip40 branch https://github.com/celo-org/celo-monorepo/blob/27a4cf7c127a667e372230e513f30ecbee22dc04/packages/sdk/contractkit/src/wrappers/OdisPayments.ts#L3-L4. It looks like they're close to identical, but it's still worth cross referencing and maybe putting up a second PR against that branch to push any changes from this PR there as well so that we don't end up with merge conflicts

Jesse-Sawa replied:
Okay - The wrapper looks pretty close but the tests are way better, I pulled them over.

alecps commented on packages/sdk/contractkit/src/wrappers/OdisPayments.test.ts:
```diff
@@ -0,0 +1,31 @@
+import { newKitFromWeb3 } from '../kit'
+import { OdisPaymentsWrapper } from './OdisPayments'
+
+testWithGanache('OdisPayments Wrapper', (web3) => {
```
As mentioned elsewhere, there's already an implementation of these tests in the odisCip40 branch. Would you mind cross-referencing? 
https://github.com/celo-org/celo-monorepo/blob/27a4cf7c127a667e372230e513f30ecbee22dc04/packages/sdk/contractkit/src/wrappers/OdisPayments.test.ts#L7

alecps commented on packages/sdk/contractkit/src/wrappers/Escrow.ts:
```diff
@@ -1,25 +1,151 @@
 export class EscrowWrapper extends BaseWrapper<Escrow> {
+  /**
+   * @notice Gets unique escrowed payments for unique payment ID.
+   * @param paymentId The ID of the payment to get.
```
Thanks for adding these docstrings! Should there be an @return here?

alecps replied:
It looks like it returns an EscrowedPayment struct

alecps commented on packages/sdk/contractkit/src/wrappers/Escrow.ts:
```diff
@@ -1,25 +1,151 @@
  */
 export class EscrowWrapper extends BaseWrapper<Escrow> {
+  /**
+   * @notice Gets unique escrowed payments for unique payment ID.
```
Nit: "Gets the unique escrowed payment for a given payment ID" since it only returns a singular object

alecps commented on packages/sdk/contractkit/src/wrappers/Escrow.ts:
```diff
@@ -1,25 +1,151 @@
+  /**
+   * @notice Gets array of all Escrowed Payments received by identifier.
+   * @param identifier The hash of an identifier of the receiver of the escrowed payment.
+   * @return An array containing all the IDs of the Escrowed Payments that were received
```
It's super confusing, but this function actually just returns a singular string. It takes in 2 params, the first is the recipients ODIS identifier and the second is the index in the escrowed payments array of the desired payment ID

Jesse-Sawa replied:
We probably don't need this as we have `getReceivedPaymentIds`

alecps commented on packages/sdk/contractkit/src/wrappers/Escrow.ts:
```diff
@@ -1,25 +1,151 @@
+  /**
+   * @notice Gets array of all Escrowed Payment IDs sent by sender.
+   * @param sender The address of the sender of the escrowed payments.
+   * @return An array containing all the IDs of the Escrowed Payments that were sent by the
```
Same as above, this actually just returns a single paymentID

alecps replied:
I might recommend just deleting receivedPaymentIds and sentPaymentIds from the wrapper in favor of the following two getters

Jesse-Sawa replied:
The [return type in the contract ](https://github.com/celo-org/celo-monorepo/blob/release/core-contracts/8/packages/protocol/contracts/identity/Escrow.sol#L342-L350)is an address array

alecps commented on packages/sdk/contractkit/src/wrappers/Escrow.ts:
```diff
@@ -1,25 +1,151 @@
   getReceivedPaymentIds = proxyCall(this.contract.methods.getReceivedPaymentIds)
 
+  /**
+   * @notice Gets array of all trusted issuers set per paymentId.
```
It looks like you may have accidentally copy pasted the wrong docstring here

Jesse-Sawa replied:
🫠

alecps commented on packages/sdk/contractkit/src/wrappers/Escrow.ts:
```diff
@@ -1,25 +1,151 @@
+    expirySeconds: number | string,
+    paymentId: string,
+    minAttestations: number | string
+  ) => CeloTransactionObject<unknown> = proxySend(this.connection, this.contract.methods.transfer)
```
should this be `<boolean>` ?

Jesse-Sawa replied:
Yes

alecps commented on packages/sdk/contractkit/src/wrappers/Escrow.test.ts:
```diff
@@ -0,0 +1,97 @@
+        1000,
+        testPaymentId,
+        1,
+        [kit.defaultAccount as string]
```
Nit: consider passing in the accounts array to test this with multiple issuers

alecps commented on packages/sdk/contractkit/src/wrappers/Escrow.test.ts:
```diff
@@ -0,0 +1,97 @@
+    const uniquePaymentIDWithdraw = withdrawKeyAddress
+    const parsedSig = await getParsedSignatureOfAddress(web3, receiver, uniquePaymentIDWithdraw)
+
+    await federatedAttestations.registerAttestationAsIssuer(identifier, receiver, TIMESTAMP).send()
```
I'd recommend using sendAndWaitForReceipt throughout to avoid potential flakiness with ganache

alecps commented on packages/sdk/contractkit/src/wrappers/Escrow.test.ts:
```diff
@@ -0,0 +1,97 @@
+    expect(senderBalanceBefore.minus(+TEN_CUSD)).toEqual(senderBalanceAfter)
+    expect(receiverBalanceBefore.plus(+TEN_CUSD)).toEqual(receiverBalanceAfter)
+  })
+})
```
Nice work!! Could you also add a test case just to make sure that withdrawal fails if the attestation isn't published by a trusted issuer?

alecps commented on packages/sdk/contractkit/src/wrappers/FederatedAttestations.ts:
```diff
@@ -1,8 +1,170 @@
+  ) => Promise<{
+    countsPerIssuer: string[]
+    identifiers: string[]
+    '0': string[]
```
curious what '0' and '1' are for

Jesse-Sawa replied:
return type from the generated file - they will be removed in the next commit

alecps commented on packages/sdk/contractkit/src/wrappers/FederatedAttestations.ts:
```diff
@@ -1,8 +1,170 @@
+  }> = proxyCall(this.contract.methods.lookupIdentifiers)
+
+  /**
+   * @notice Helper function for lookupAttestations to calculate the total number of
```
It looks like you accidentally copied the docstring for getNumAttestations

Jesse-Sawa replied:
big flop

alecps commented on packages/sdk/contractkit/src/wrappers/FederatedAttestations.ts:
```diff
@@ -1,8 +1,170 @@
+   * @param account Address of the account mapped to the identifier
+   * @dev Throws if sender is not the issuer, signer, or account
+   */
+  revokedAttestations: (
```
Could you change this to "revokeAttestation"? This could be confused as the getter for the revokedAttestations array

alecps commented on packages/sdk/contractkit/src/wrappers/FederatedAttestations.ts:
```diff
@@ -1,8 +1,170 @@
+    v: string | number,
+    r: string | number[],
+    s: string | number[]
+  ) => CeloTransactionObject<void> = proxySend(
```
This is a view function so you can just use proxyCall

alecps commented on packages/sdk/contractkit/src/wrappers/FederatedAttestations.ts:
```diff
@@ -1,8 +1,170 @@
+    account: Address,
+    signer: Address,
+    issuedOn: number | string
+  ) => CeloTransactionObject<string> = proxySend(
```
This is a 'pure' function so you can just use proxyCall

alecps commented on packages/sdk/contractkit/src/wrappers/FederatedAttestations.test.ts:
```diff
@@ -16,11 +12,104 @@ testWithGanache('FederatedAttestations Wrapper', (web3) => {
-      federatedAttestations = await kit.contracts.getFederatedAttestations()
-      expect(federatedAttestations)
+  it('no identifiers should exist if none were registered', async () => {
+    federatedAttestations = await kit.contracts.getFederatedAttestations()
```
Is there a reason you have this here instead of in the beforeAll hook?

alecps commented on packages/sdk/contractkit/src/wrappers/FederatedAttestations.test.ts:
```diff
@@ -16,11 +12,104 @@ testWithGanache('FederatedAttestations Wrapper', (web3) => {
+      kit.defaultAccount as string,
+    ])
+
+    expect(attestations[1]).toEqual([testAccount])
```
This might be clearer if you write it to match the entire attestations and identifiers objects. I don't completely understand how the indexing works here

Jesse-Sawa replied:
Yeah solid - I have to change the return types on the wrappers

alecps commented on packages/sdk/contractkit/src/wrappers/FederatedAttestations.test.ts:
```diff
@@ -16,11 +12,104 @@ testWithGanache('FederatedAttestations Wrapper', (web3) => {
+      [kit.defaultAccount as string]
+    )
+
+    expect(attestationsAfterRevocation).toEqual({
```
Could you check identifiersAfterRevocation as well (consider renaming test or splitting this check out to match pattern in first two tests)

alecps commented on packages/sdk/contractkit/src/wrappers/FederatedAttestations.test.ts:
```diff
@@ -16,11 +12,104 @@ testWithGanache('FederatedAttestations Wrapper', (web3) => {
+      countsPerIssuer: ['0'],
+      issuedOns: [],
+      publishedOns: [],
+      signers: [],
```
It would be great if these tests could cover registerAttestation() and batchRevokeAttestations() so that we have some code showing how to invoke these functions as we begin working on the rest of the sdk. Does that sound like a manageable addition to this PR or would you prefer to tackle / test this elsewhere in the sdk?

Jesse-Sawa replied:
[Created a ticket ](https://github.com/celo-org/celo-monorepo/issues/9928) to add the remaining test

alecps reviewed (changes requested):
Thanks for cranking this out!! Looks really good so far

isabellewei commented on packages/sdk/contractkit/src/wrappers/FederatedAttestations.ts:
```diff
@@ -1,8 +1,170 @@
-// TODO ASv2 -- add params or delete config if there are none
-// & delete other commented-out usages of this Config in CK
-// export interface FederatedAttestationsConfig {}
+type Address = string
```
you can import this type from `@celo/connect`, example from the Accounts wrapper: https://github.com/celo-org/celo-monorepo/blob/1acca063339adaac710341279d3ccfaa029e4b0a/packages/sdk/contractkit/src/wrappers/Accounts.ts#L2

isabellewei commented on packages/sdk/contractkit/src/wrappers/FederatedAttestations.ts:
```diff
@@ -1,8 +1,170 @@
-// & delete other commented-out usages of this Config in CK
-// export interface FederatedAttestationsConfig {}
+type Address = string
+type Identifier = string | number[]
```
the identifier is always a string! I feel like declaring a type variable for it isn't necessary, but up to you

isabellewei commented on packages/sdk/contractkit/src/wrappers/FederatedAttestations.ts:
```diff
@@ -1,8 +1,170 @@
+    signers: string[]
+    issuedOns: string[]
+    publishedOns: string[]
+    '0': string[]
```
what are these number fields for?

Jesse-Sawa replied:
These were the generate parameters for the solidity file - they have been removed

isabellewei commented on packages/sdk/contractkit/src/wrappers/FederatedAttestations.ts:
```diff
@@ -1,8 +1,170 @@
+    account: Address,
+    signer: Address,
+    issuedOn: number | string,
+    v: number | string,
```
it would be nice if we could pass in a `Signature` object here instead :) example: https://github.com/celo-org/celo-monorepo/blob/1acca063339adaac710341279d3ccfaa029e4b0a/packages/sdk/contractkit/src/wrappers/Accounts.ts#L162-L175

Jesse-Sawa replied:
After implementing, I've noticed you can't call `.send()` directly but instead have to create the Object then submit the transaction. 

In effort of following the same pattern for all functions in this wrapper I would suggest to revert back to the individual parameters and rely on the jsdoc to explain

alecps replied:
Good point Izzy. I think we should definitely strive to use the Signature type when we invoke this function, and when I first read your comment I agreed completely. I also see Jesse's point though on how keeping it this way allows us to pass the parameters through to proxySend() without any processing, which is a little cleaner IMO so I might also be in favor of keeping it this way and parsing v,r,s out one level up. Wdyt?

isabellewei replied:
sounds good! We can use the Signature type on the sdk side instead :)

isabellewei commented on packages/sdk/contractkit/src/wrappers/FederatedAttestations.ts:
```diff
@@ -1,8 +1,170 @@
+    account: Address,
+    signer: Address,
+    issuedOn: string | number,
+    v: string | number,
```
ditto for `Signature` object and `issuedOn` always being a number

Jesse-Sawa committed:
FederatedAttestations wrapper formatting, return types, proxyCall fixes
Escrow wrapper test for failed withdrawals
Replace send() with sendAndAwaitReceipt
Escrow test Pass in accounts array instead of single issuer
OdisPayments use existing tests from OdisCip40

isabellewei commented on packages/sdk/contractkit/src/wrappers/FederatedAttestations.ts:
```diff
@@ -1,8 +1,170 @@
+    issuer: Address,
+    account: Address,
+    signer: Address,
+    issuedOn: number | string,
```
this is always a number!

isabellewei commented on packages/sdk/contractkit/src/wrappers/Escrow.ts:
```diff
@@ -1,25 +1,151 @@
+    value: number | string,
+    expirySeconds: number | string,
+    paymentId: string,
+    minAttestations: number | string
```
`value`, `expirySeconds`, and `minAttestations` are always numbers, and `paymentId` is an address

isabellewei commented on packages/sdk/contractkit/src/wrappers/OdisPayments.ts:
```diff
@@ -0,0 +1,26 @@
+   * @param value The amount in cUSD to pay.
+   * @dev Throws if cUSD transfer fails.
+   */
+  payInCUSD: (account: Address, value: number | string) => CeloTransactionObject<void> = proxySend(
```
`value` is a number always

Jesse-Sawa replied:
So it can be passed as either a number or a string here. We can choose to constrain it to a number but I don't really see the benefit of that

isabellewei commented on packages/sdk/contractkit/src/wrappers/OdisPayments.ts:
```diff
@@ -0,0 +1,26 @@
+export class OdisPaymentsWrapper extends BaseWrapper<OdisPayments> {
+  /**
+   * @notice Fetches total amount sent (all-time) for given account to odisPayments
+   * @param account The account to fetch all-time quota for
```
nit: replace 'quota' with 'funds' or 'payments'. we decided on using quota to refer to the number of queries available to the user and payments to refer to the money they've paid towards obtaining quota, to make things a little less confusing

Jesse-Sawa replied:
yes

isabellewei commented on packages/sdk/contractkit/src/wrappers/Escrow.test.ts:
```diff
@@ -0,0 +1,126 @@
+    expect(senderBalanceBefore.minus(+TEN_CUSD)).toEqual(senderBalanceAfter)
+    expect(receiverBalanceBefore.plus(+TEN_CUSD)).toEqual(receiverBalanceAfter)
+  })
+  it('withdraw should revert if attestation is not registered', async () => {
```
would also be great to do one where withdraw fails when attestation is registered by an issuer not on the trusted issuer list :)

Jesse-Sawa replied:
Ah yes def agree that this is a good test although, I do think it doesn't fit into the scope of this suite. While we do test different flows, the aim here is to test the wrapper functionality as opposed to alternate flows/failure paths

alecps replied:
I see both sides here. I agree that this case probably should be tested at the protocol level, but it looks like it's also missing there. @Jesse-Sawa maybe you could just put a TODO(ASv2) statement on this line to help us remember to add it? https://github.com/celo-org/celo-monorepo/blob/1522250ff90d0ab1d8afd328d32ac15ce160b1e5/packages/protocol/test/identity/escrow.ts#L1121

Jesse-Sawa replied:
Sure thing - I'll actually add it to the ["Add registerAttestation test" issue](https://github.com/celo-org/celo-monorepo/issues/9928) and knock them both out in a PR

Jesse-Sawa committed:
minor type fix

Jesse-Sawa committed:
given the pattern with existing tools, value should be able to accept a string

alecps commented on packages/sdk/contractkit/src/wrappers/Escrow.test.ts:
```diff
@@ -0,0 +1,126 @@
+        1,
+        accounts
+      )
+      .send({ from: sender })
```
nit: sendAndWaitForReceipt

alecps commented on packages/sdk/contractkit/src/wrappers/Escrow.test.ts:
```diff
@@ -0,0 +1,126 @@
+        1,
+        accounts
+      )
+      .send({ from: sender })
```
nit: sendAndWaitForReceipt

alecps reviewed (approved):
Looks awesome!! Just a couple nits about missing sendAndWaitForReceipt's.

The helper function you were probably looking for to submit registerAttestation is here https://github.com/celo-org/celo-monorepo/blob/1522250ff90d0ab1d8afd328d32ac15ce160b1e5/packages/protocol/lib/fed-attestations-utils.ts#L44, which is admittedly not a great a place long term (I think the plan was to move these helpers into the SDK). I see you've already created a ticket to add a registerAttestation test which is great 🚀  You might also want to create a ticket to move the attestation signing logic into the identity sdk so we don't forget that's on the roadmap

Jesse-Sawa committed:
replace send() with sendAndWaitForReceipt()

Jesse-Sawa committed:
Merge branch 'master' into identity/asv2-contract-wrappers

Jesse-Sawa commented:
> Looks awesome!! Just a couple nits about missing sendAndWaitForReceipt's.
> 
> The helper function you were probably looking for to submit registerAttestation is here
> 
> https://github.com/celo-org/celo-monorepo/blob/1522250ff90d0ab1d8afd328d32ac15ce160b1e5/packages/protocol/lib/fed-attestations-utils.ts#L44
> 
> , which is admittedly not a great a place long term (I think the plan was to move these helpers into the SDK). I see you've already created a ticket to add a registerAttestation test which is great 🚀 You might also want to create a ticket to move the attestation signing logic into the identity sdk so we don't forget that's on the roadmap

Thank you for the feedback @isabellewei & @alecps! 
I created two tickets to follow up on this PR: 
* [Add tests ](https://github.com/celo-org/celo-monorepo/issues/9928) 
* [Move the attestation signing logic](https://github.com/celo-org/celo-monorepo/issues/9929) 

Also replaced those lingering `send()`s

isabellewei commented:
closes #9448
